File: problems/java/MeetingScheduler/src/services/NotificationService.java
package services;

import enums.NotificationType;
import models.Meeting;
import models.Notification;
import models.User;

import java.time.LocalDateTime;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

public class NotificationService {
    private static final AtomicInteger notificationId = new AtomicInteger(1);

    public void sendNotifications(NotificationType type, List<User> participants, Meeting meeting) {
        Notification notification = new Notification(
                notificationId.getAndIncrement(),
                type,
                type.getSubjectPrefix() + ": " + meeting.getSubject(),
                LocalDateTime.now());
        for (User participant : participants) {
            notification.send(participant, meeting);
        }
    }
}

File: problems/python/MeetingScheduler/src/services/notification_service.py
from datetime import datetime
from itertools import count
from threading import Lock

from enums.notification_type import NotificationType
from models.meeting import Meeting
from models.notification import Notification
from models.user import User


class NotificationService:
    _ids = count(1)
    _id_lock = Lock()

    def send_notifications(
        self, notification_type: NotificationType, participants: list[User], meeting: Meeting
    ) -> None:
        with NotificationService._id_lock:
            notification_id = next(NotificationService._ids)

        notification = Notification(
            notification_id=notification_id,
            notification_type=notification_type,
            content=f"{notification_type.subject_prefix}: {meeting.subject}",
            creation_date=datetime.now(),
        )
        for participant in participants:
            notification.send(participant, meeting)
